### ISSUE TITLE
Master is calling ADD_PROCS on all procs

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#4440]
After studying the connectivity problems on master when performing async modex, I have discovered that someone modified MPI_Init so that we now call add_procs on _every_ process regardless of the setting of mpi_add_procs_cutoff.

As folks probably recall, we had agreed as a community not to call add_procs on a peer until first communication. We even set the default cutoff value to 0 as we didn't see any launch performance negatives to that setting.

I'm not very familiar with the code here, but I see a call to PML_ADD_PROCS in ompi_mpi_init.c:837 that isn't protected by the cutoff. It is this call that is causing the TCP BTL to add every process.

Can someone please take a look and see what needs to be fixed?

**Comment by bosilca** [comment]
Someone is a bad boy ! I'll take a look. Any hint on how to consistently reproduce it?

**Comment by rhc54** [comment]
All you need is two nodes and then execute:
```
$ mpirun -mca pmix_base_async_modex 1 -mca pmix_base_collect_data 0 -npernode 2 ./mpi_no_op
```
from the orte/test/mpi directory (just MPI_Init/MPI_Finalize, so no communication). The combination of those two params drops us directly thru the modex without executing any barrier. What happens then is that add procs gets called and believes that there is no way for two procs on the same node to communicate, and so you fail.

There is clearly an error in PMIx here, but a little debug output shows that the btl is requesting modex info for every process.

**Comment by rhc54** [comment]
@bwbarrett @mohanasudhan I believe you folks were working on the reachability code in the TCP BTL? I'm wondering if somehow during that effort the code was changed so that it now attempts to determine reachability to every other process in the job during init? It is supposed to only determine reachability to a given process upon first communication with that proc.

This does two things: (a) it means we only store ompi_proc_t structs for those procs we communicate with, and (b) we only pull modex info when we actually need it. Both are critical for scalability.

**Comment by bwbarrett** [comment]
@rhc54, nope, we never committed the changes to the TCP BTL.  Just the util code and the reachability framework (which gets its data from the BTL/whatever, not from modex).  For once, our delay in finishing that code comes in handy :)

**Comment by bosilca** [comment]
@rhc54 I cant confirm your findings. I have tried the UCX PML and it works just fine (up to 128 processes over 8 nodes). Also PML OB1 works all the time with openib, vader and self (same 128 processes), but crashed when TCP is on (but this is weird as it crashes in during vader memory allocation). I could try to take a look at the TCP BTL, but as you mentionned that @bwbarrett is on it, I will let him finish his work.

**Comment by rhc54** [comment]
Sigh - I believe I did say that it was something to do with the TCP BTL, didn't I? And my takeaway from @bwbarrett's comment is that they aren't working on it, so I guess nobody is for now.

I guess I'll just go ahead and address the PMIx issue (as I have an independent reproducer there) and leave the OMPI scaling issues to the OMPI team. I suspect this is the same as #4277, so I'll close this one.

**Comment by bwbarrett** [comment]
We'll take a look at the TCP BTL; I didn't say we weren't working on it, only that we didn't touch the wireup path yet.

**Comment by bwbarrett** [comment]
@bosilca, can you file an issue about your crash?

**Comment by rhc54** [comment]
@bwbarrett thanks - sorry I misunderstood your earlier comment. You might also keep in mind #4264 as it may be related to the crash.

**Comment by ggouaillardet** [comment]
@rhc54 here is the code before invoking `add_procs`

```c
    /* some btls/mtls require we call add_procs with all procs in the job.
     * since the btls/mtls have no visibility here it is up to the pml to
     * convey this requirement */
    if (mca_pml_base_requires_world ()) {
        ...
    } else {
    }
```

your observation suggests that `btl/tcp` requires `add_procs` is invoked on all procs.
not sure whether this is a bug / feature / room for improvement.

@bosilca @bwbarret any recollection on why `btl/tcp` seems to require this ?

**Comment by ggouaillardet** [comment]
@rhc54 are you sure `add_procs()` is invoked on **all** the MPI tasks in `ompi_mpi_init()` ?

i used the patch below, and at this stage, `add_procs()` is only invoked on the **local** MPI tasks (e.g. the tasks running on the same node)

for example
```
n0$ mpirun --host n0:4,n1:4 -mca pmix_base_async_modex 1 -mca pmix_base_collect_data 0 -npernode 2 ./ring_c
[n0:64974] rank 0 does NOT require world 2 procs : 0 1
[n0:64975] rank 1 does NOT require world 2 procs : 0 1
[n1:38464] rank 2 does NOT require world 2 procs : 2 3
[n1:38465] rank 3 does NOT require world 2 procs : 2 3
```

```patch
diff --git a/ompi/runtime/ompi_mpi_init.c b/ompi/runtime/ompi_mpi_init.c
index f49ac00..9f0f0ea 100644
--- a/ompi/runtime/ompi_mpi_init.c
+++ b/ompi/runtime/ompi_mpi_init.c
@@ -821,17 +821,27 @@ int ompi_mpi_init(int argc, char **argv, int requested, int *provided)
      * since the btls/mtls have no visibility here it is up to the pml to
      * convey this requirement */
     if (mca_pml_base_requires_world ()) {
+        char buf[1024];
+        char * p = buf;
         if (NULL == (procs = ompi_proc_world (&nprocs))) {
             error = "ompi_proc_get_allocated () failed";
             goto error;
         }
+        p += sprintf(p, "rank %d DOES requires world %d procs :", OMPI_PROC_MY_NAME->vpid, nprocs);
+        for (int i=0; i< nprocs; i++) p += sprintf(p, " %d", procs[i]->super.proc_name.vpid);
+        OPAL_OUTPUT((0, buf));
     } else {
+        char buf[1024];
+        char * p = buf;
         /* add all allocated ompi_proc_t's to PML (below the add_procs limit this
          * behaves identically to ompi_proc_world ()) */
         if (NULL == (procs = ompi_proc_get_allocated (&nprocs))) {
             error = "ompi_proc_get_allocated () failed";
             goto error;
         }
+        p += sprintf(p, "rank %d does NOT require world %d procs :", OMPI_PROC_MY_NAME->vpid, nprocs);
+        for (int i=0; i< nprocs; i++) p += sprintf(p, " %d", procs[i]->super.proc_name.vpid);
+        OPAL_OUTPUT((0, buf));
     }
     ret = MCA_PML_CALL(add_procs(procs, nprocs));
     free(procs);
```

**Comment by rhc54** [comment]
I can't speak to your patch, but if you use those two mca params, and you watch the modex data requests, you definitely see a request for the btl.tcp key by each process for every process

**Comment by ggouaillardet** [comment]
where do you "plug to" in order to see these requests ?
i added some `printf()` in `PMIX_Get_nb()` and could not see any requests for any remote MPI task.

**Comment by ggouaillardet** [comment]
i did some more checks, and in `ompi_mpi_init()` there are two paths
 1. `mca_pml_base_requires_world()` is `true`, this has been ordered by the `pml`, which was ordered by the `btl` or `mtl`. so far, `btl/portals4` (under some conditions), `btl/usnic`, `mxm/mtl_mxm` and `mtl/portals4` (under some conditions) can order this. In this case, `add_proc`  in invoked on __all__ MPI tasks.
2. `mca_pml_base_require_world()` is false, and in this case `add_proc` is invoked on the MPI tasks returned by  `ompi_proc_get_allocated()`. this is all the procs in `ompi_proc_list`, and as far as i can see, it is populated with the local task and all those returned by `OPAL_PMIX_LOCAL_PEERS`

bottom line, and unless a specific `btl` or `mtl` is used (or is causing undesired side effect), or `OPAL_PMIX_LOCAL_PEERS` is busted and return also non local peers, i do not see how `add_proc` can be invoked on __non local__ tasks in `MPI_Init()`

**Comment by rhc54** [comment]
I think we have tracked this down now and have a fix in the works.
